workers/unity/Packages/io.improbable.gdk.core/Utility/EntitySnapshot.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            var handler = new SchemaEntitySerializer(inObj, components);
+
+            foreach (var pair in components)
+            {
+                Dynamic.ForComponent(pair.Key, handler);
+            }
+        }
+    }
+
+    internal struct QueryComponentDataHandler : Dynamic.IHandler
+    {
+        public Dictionary<uint, ISpatialComponentSnapshot> Components;
+
+        private readonly Entity entitySnapshot;
+
+        public QueryComponentDataHandler(Entity entitySnapshot)
+        {
+            Components = new Dictionary<uint, ISpatialComponentSnapshot>();
+            this.entitySnapshot = entitySnapshot;
+        }
+
+        public void Accept<TData, TUpdate, TSnapshot>(uint componentId, Dynamic.VTable<TData, TUpdate, TSnapshot> vtable)
+            where TData : struct, ISpatialComponentData
+            where TUpdate : struct, ISpatialComponentUpdate
+            where TSnapshot : struct, ISpatialComponentSnapshot
+        {
+            var schemaObject = entitySnapshot.Get(componentId).Value;
+            Components.Add(componentId, vtable.DeserializeSnapshot(schemaObject));
+        }
+    }
+
+    internal struct SchemaEntityDeserializer : Dynamic.IHandler
+    {
+        public readonly Dictionary<uint, ISpatialComponentSnapshot> Components;
+
+        private SchemaObject entityObject;
+
+        public SchemaEntityDeserializer(SchemaObject entityObject)
+        {
+            Components = new Dictionary<uint, ISpatialComponentSnapshot>();
+            this.entityObject = entityObject;
+        }
+
+        public void Accept<TData, TUpdate, TSnapshot>(uint componentId, Dynamic.VTable<TData, TUpdate, TSnapshot> vtable)
+            where TData : struct, ISpatialComponentData
+            where TUpdate : struct, ISpatialComponentUpdate
+            where TSnapshot : struct, ISpatialComponentSnapshot
+        {
+            var componentObject = entityObject.GetObject(componentId);
+            Components[componentId] = vtable.DeserializeSnapshotRaw(componentObject);
+        }
+    }
+
+    internal struct SchemaEntitySerializer : Dynamic.IHandler
+    {
+        private Dictionary<uint, ISpatialComponentSnapshot> componentSnapshots;
+        private SchemaObject targetObject;
+
+        public SchemaEntitySerializer(SchemaObject targetObject, Dictionary<uint, ISpatialComponentSnapshot> componentSnapshots)
+        {
+            this.targetObject = targetObject;
+            this.componentSnapshots = componentSnapshots;
+        }
+
+        public void Accept<TData, TUpdate, TSnapshot>(uint componentId, Dynamic.VTable<TData, TUpdate, TSnapshot> vtable)
+            where TData : struct, ISpatialComponentData
+            where TUpdate : struct, ISpatialComponentUpdate
+            where TSnapshot : struct, ISpatialComponentSnapshot
+        {
+            // Okay to grab the value directly, we only call this for snapshots that actually exist.
```

Comment:
how do we know/guarantee that this will only be called for snapshots that exist?

Reply:
We call this directly in the in ` loop where we are iterating over the same dictionary. Its a little awkward but it means we don't waste time checking for components that aren't there. 

https://github.com/spatialos/gdk-for-unity/pull/1053/files/86df0c90b7bb41a9af8cc783902f308bc1e864e2#diff-fa3492c073faead7e4b8822645ed07adR88